bootstrap/012_task_log.sql

```diff
@@ -152,3 +160,47 @@ CREATE OR REPLACE VIEW ADMIN.TASK_LOG_HISTORY AS SELECT * FROM INTERNAL.TASK_LOG
 
 -- Remove view from old location
 DROP VIEW IF EXISTS REPORTING.TASK_LOG_HISTORY;
+
+CREATE OR REPLACE PROCEDURE INTERNAL.CLOSE_LOG_FOR_TASK(task_log_table text, query_history_table text, task_name text, object_name text)
+    RETURNS OBJECT
+    LANGUAGE SQL
+    Comment = 'Close any outstanding task rows that have not been updated since the given timestamp. If a query_id is present, the query_history row is set as the output'
+AS
+BEGIN
+    let outstanding_rows resultset := (select task_start, task_run_id, query_id from IDENTIFIER(:task_log_table) where success is null and task_name = :task_name and object_name = :object_name);
+    let cur cursor for outstanding_rows;
+    let rows_closed integer := 0;
+    for r in cur do
+        rows_closed := rows_closed + 1;
+        let task_start text := r.task_start;
+        let task_run_id text := r.task_run_id;
+        let query_id text := r.query_id;
+        let output object;
+
+        if (query_id is not null) then
+            -- Try to look up the query history details if we have a query_id
+            output := (select object_construct(*) from IDENTIFIER(:query_history_table) where query_id = :query_id limit 1);
+        end if;
+
+        // Always include a 'force_closed' key to indicate rows closed by this operation for post-hoc analysis
+        if (output is null) THEN
+            let msg text := (SELECT 'Could not find query history for task=' || :task_name || ', object=' || :object_name || ', task_run_id=' || :task_run_id || ', query_id=' || COALESCE(:query_id, 'NULL'));
+            SYSTEM$LOG_INFO(:msg);
+            output := OBJECT_CONSTRUCT('FORCE_CLOSED', TRUE, 'details', :msg);
+        else
+            output := OBJECT_INSERT(:output, 'FORCE_CLOSED', TRUE);
+        end if;
+
+        -- Mark the row as having failed.
+        update IDENTIFIER(:task_log_table) set success = FALSE, output = :output, task_finish = current_timestamp()::TIMESTAMP_NTZ
+            where task_start = :task_start and task_run_id = :task_run_id and task_name = :task_name and object_name = :object_name and success is null;
+    end for;
+
+    let o object := OBJECT_CONSTRUCT('success', True, 'rows_closed', :rows_closed);
+    SYSTEM$LOG_INFO(:o);
+    return o;
+EXCEPTION
+    WHEN OTHER THEN
+        SYSTEM$LOG_ERROR(OBJECT_CONSTRUCT('error', 'Exception trying to close outstanding task rows.', 'SQLCODE', :sqlcode, 'SQLERRM', :sqlerrm, 'SQLSTATE', :sqlstate));
+        return OBJECT_CONSTRUCT('SQLCODE', :sqlcode, 'SQLERRM', :sqlerrm, 'SQLSTATE', :sqlstate);
+END;
```

Comment:
The procedural version seems heavy and hard for me to parse. I've got:

```
update <insert update here> from
with foo as (
select current_timestamp as task_start, 'foo' as task_run_id, '01b2438c-0904-46df-0062-5003077b8cfe' as query_id,
'Could not find query history for task=' || 'task_name' || ', object=' || 'objecT_name' || ', task_run_id=' || task_run_id || ', query_id=' || COALESCE(query_id, 'NULL') as msg, OBJECT_CONSTRUCT('FORCE_CLOSED', TRUE) output
), bar as (
select object_construct(q.*) as qh, task_start, task_run_id, msg, output from snowflake.account_usage.query_history q inner join foo on q.query_id = foo.query_id
)
select case when qh is null then objecT_insert(output, 'details', msg) else  dryan.internal.merge_objects(output, qh) end as output from bar;
```

working. I think the changes would be:
* use right outer join
* i think more things need to be passed through the CTE
* finish the update statement

Reply:
Thanks for the snippet. I worry that this is going to be slow as molasses because querying the SecureView is going to prevent the predicate pushdown from happening.

When I run your example (with a constant value for the query_id), I can see that only one row is returned from query_history in a few seconds. However, if I replace `foo` with a query that reads a query_id from a table, then the query profile shows scanning thousands of rows of query_history, running for over a minute without completing.

Two options I'm looking at:
* I can try to clean up the creation of the `output` to be a little less ugly, and we continue to go down the path I started to on this PR
* We stop trying to do the query_history aggregation at task execution time, and create a 2nd view in the native app which does the join against our materialized view query history. Something like
```
create view admin.materialization_failures as 
with failures as (
  select * from admin.task_log_history where success is null output:forced_closed = true
) select *, object_construct(eqh) as query_history_details from failures
left join reporting.enriched_query_history eqh on failures.query_id = eqh.query_id;
```

Reply:
I ended up doing the first option from my last comment as it was most straightforward to implement without writing an entirely new piece to this PR.

I'm not particularly surprised about the secure view performance being bad for the join case (like you originally suggested) and a `from query_history where query_id in (select query_id from tasks_needing_closed)`, but it's frustrating none the less.

The current approach in here runs in ~4seconds which feels reasonable.

Reply:
Is it that the predicate isn't being pushed down or is it that we don't have a good predicate? Is it still slow if we limit by time? eg `from query_history where query_id in (select query_id from tasks_needing_closed) and start_time > current_timestamp - interval '1 day'`

Reply:
> Is it that the predicate isn't being pushed down or is it that we don't have a good predicate?

Well, I can't say that I know for certain, but it feels like the optimizer doing something in one case and not the other (e.g. constant string query_id -> fast and reads one row. query_id taken from CTE expression -> slow, reading many rows).

> Is it still slow if we limit by time

I'll check.

Reply:
> Is it still slow if we limit by time

yeah, looks like filtering on time does greatly help.